README.md
# ServerKit

[Invite link](https://discord.com/oauth2/authorize?client_id=828582617254461481&permissions=2587094358&scope=bot%20applications.commands)

ServerKit is a Discord bot that combats bot features that are locked by paywall and reduce the number of bots needed in a server.

## Table of contents

* [Technologies used](#technologies-used)
* [Setup](#setup)
* [Features](#features)
* [Roadmap](#roadmap)
* [Support](#support)
* [Self-hosting and using the bot's code](#self-hosting-or-using-the-bots-code-in-general)
* [Contributing](#contributing)
* [Project status](#project-status)
* [Inspiration](#Inspiration)
* [License](#license)

## Technologies used

| Technology | Description | Link |
| ----------- | ----------- | ----------- |
| Javascript | 'JavaScript (JS) is a lightweight, interpreted, or just-in-time compiled programming language with first-class functions." - Mozilla MDN | None |
| Typescript | Builds on JavaScript, one of the world’s most used tools, by adding static type definitions. | https://www.typescriptlang.org/ |
| discord.js | discord.js is a powerful node.js module that allows you to interact with the Discord API very easily. | https://discord.js.org/ |
| discord.ts | Is an extension of discord.js | https://github.com/OwenCalvin/discord.ts |
| FFmpeg | A complete, cross-platform solution to record, convert and stream audio and video. | https://www.ffmpeg.org/ |
| youtube-dl | Command-line program to download videos from YouTube.com and other video sites. | https://github.com/ytdl-org/youtube-dl |

## Setup

1. Install latest version of above list.
2. Create a .env file with the contents:
```
PRODUCTION=False
PRODUCTIONTOKEN=(Leave blank)
DEVTOKEN=(Your bot token)
```
3. Start `main.py`

## Features

### Open Source (and being Libre)
Being open source benefits the users, everyone can submit bugs, check code, or even pull request to make the bot work better.

### Free, forever
No patreon bs.

### Other features
* Slash Command Support
* Moderation commands
* Music commands
* Fun commands

## Roadmap
- More fun features (View memes and create memes)
- Welcomer features
- More moderation features (Logging, etc.)

## Support

Join the Support server on [Discord](https://discord.gg/CqRkKpZR), or if the bug impacts all users, submit a issue here.

## Self-hosting (or using the bot's code in general)

You can only self-host this bot if you want to test your changes.
As per the license requires, PLEASE DON'T TAKE THE BOT'S CODE AS YOUR OWN.

## Contributing

I am ABSOLUTELY open to any contributions, please make sure you state what changed.

## Project status

Project is in progress and allow everyone to use and test.

## Inspiration

Thanks to all major Discord bots out there to make me develop this bot! Specially those bots that it's features are locked by paywalls.

## License
Up until version 0.3.5.1, it is in [GPL license](https://www.gnu.org/licenses/gpl-3.0.html), after 0.3.5.1, it will be switched to [AGPL License](https://www.gnu.org/licenses/agpl-3.0.html) since Discord Bots counts as server software (duh.)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project is adhere to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Changed
- License changed from GPLv3 to AGPLv3.
- Switched to a new api wrapper (discord.js).
- Updated the README and CHANGELOG since the bot is added to the curated list of open source Discord bots.

## [Alpha 0.3.5.1] - 2021-04-12
### Fixed
- One line in Music command (wait(1)) caused "This interaction failed" message but plays music while using /play.

## Alpha 0.3.5 - 2021-04-12
### Changed
- Specify that Spotify links works with play command.

### Fixed
- Fix for "This interaction failed" message while using slash commands.

## Alpha 0.3.4 - 2021-04-11
### Fixed
- Fix for an extra ) in the Ban command.

## Alpha 0.3.3 - 2021-04-11
### Fixed
- Fix slash command options because of me forgot to import `create_option`.

## Alpha 0.3.2 - 2021-04-11
### Changed
- Include option explainations into slash commands.

### Fixed
- Update pulling reason text from user's message to prevent error.

## Alpha 0.3.1 - 2021-04-11
### Changed
- Includes slash command in help command and bot's status.

### Fixed
- Fixed moderation command due to a misnaming in the async functions.

## Alpha 0.3 - 2021-04-11
### Added
- Slash command support.
- @someone command added. (throwback to Discord April Fools 2018)

### Changed
- Update invite link to support slash command and replace Administrstor permission with the required ones.

### Deprecated
- Move the contents of changelog and info command to Github and Help command respectively, command still exists.

### Fixed
- Fixed embed not showing up in play, stop, pause commands while user isn't in a voice channel.
- Attempting to fix 403 Error command
- Removed useless options in youtube-dl

## Alpha 0.2.3 - 2021-04-09
### Added
- More fun commands (Eightball, dice)
- Moderation commands (Ban, Kick)

### Changed
- Moved many commands into a cog. (info, changelog, help and ping are in info cog)

## Alpha 0.2.2 - 2021-04-08?
### Changed
- New branding change.

### Fixed
- Bug fixes.

## Alpha 0.2.1 - 2021-04-08?
### Fixed
- Bug fixes.

## Alpha 0.2 - 2021-04-08?
### Added
- Music commands (join, leave, play, pause, stop)
- Changelog command

### Fixed
- Bug Fixes

## Alpha 0.1.1.4 - 2021-04-07?
### Fixed
- Bug Fixes

[Unreleased]:
https://github.com/BugGlitchy64/ServerKit/compare/v0.3.5.1-alpha...HEAD
[Alpha 0.3.5.1]:
https://github.com/BugGlitchy64/ServerKit/releases/tag/v0.3.5.1-alpha


cogs/info.py
# Copyright (C) 2021 BugGlitchy64
# 
# This file is part of ServerKit.
# 
# ServerKit is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# ServerKit is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with ServerKit.  If not, see <http://www.gnu.org/licenses/>.

import discord
from discord.ext import commands
from discord_slash import cog_ext
from discord_slash.utils.manage_commands import create_option

class info(commands.Cog):

    def __init__(self, client):
        self.client = client
        client.remove_command('help')

    @commands.Cog.listener()
    async def on_ready(self):
        print("[DEBUG] Info is OK!")

    async def info(self, ctx, normalCommand):
        embed = discord.Embed(title = 'ServerKit', color = self.client.color)
        embed.description = f"This command has been moved to {self.client.command_prefix}help"
        if normalCommand:
            await ctx.send(embed = embed, reference = ctx.message)
        else:
            await ctx.send(embed = embed)

    async def help(self, ctx, normalCommand, arg=None):
        cogs = [c for c in self.client.cogs.keys()]
        if arg == None:
            embed = discord.Embed(title = 'ℹ️ Important commands', color = self.client.color)
            embed.description = f'Here are a list of some important commands in the bot, prefix are `{self.client.command_prefix}` and `/`.\nType `{self.client.command_prefix}help (command)` for more info.\nType `{self.client.command_prefix}help 2` for miscellaneous commands.'
            embed.add_field(name = '🔨 Moderation', value = '`ban` `kick` `mute*` `warn*`', inline = True)
            embed.add_field(name = '🎵 Music', value = '`join` `play` `stop` `skip*` `voteskip*` `pause` `queue*`', inline = True)
            embed.add_field(name = 'Links', value = '[Support Server](https://discord.gg/YA5pTheC3A) [Invite link](https://discord.com/api/oauth2/authorize?client_id=828582617254461481&permissions=2587094358&scope=bot%20applications.commands) [Github](https://github.com/BugGlitchy64/ServerKit)', inline = False)
            embed.set_footer(text = 'You are in page 1/2. (* ones are in construction)')
        elif arg == "2":
            embed = discord.Embed(title = 'ℹ️ Miscellaneous commands', color = self.client.color)
            embed.description = f'Prefix are `{self.client.command_prefix}` or `/`.\nTo return to the important commands, type `{self.client.command_prefix}help`.'
            embed.add_field(name = 'ℹ️ Information', value = '`help` `info` `ping` `changelog`', inline = True)
            embed.add_field(name = '😂 Fun', value = '`eightball` `dice` `thoughts` `someone`', inline = True)
            embed.add_field(name = 'Links', value = '[Support Server](https://discord.gg/YA5pTheC3A) [Invite link](https://discord.com/api/oauth2/authorize?client_id=828582617254461481&permissions=2587094358&scope=bot%20applications.commands) [Github](https://github.com/BugGlitchy64/ServerKit)', inline = False)
            embed.set_footer(text = 'You are in page 2/2. (* ones are in construction)')
        else:
            noCommand = True
            lowArg = arg.lower()
            for cog in cogs:
                for command in self.client.get_cog(cog).walk_commands():
                    if lowArg == command.name:
                        embed = discord.Embed(title = f'{self.client.command_prefix}{command.name} (or /{command.name})', color = self.client.color)
                        embed.description = f'`{command.description}`'
                        embed.add_field(name = 'Category', value = command.usage, inline = True)
                        embed.add_field(name = 'Alias(es)', value = command.aliases, inline = True)
                        noCommand = False 
                        break
            if noCommand == True:    
                embed = discord.Embed(title = "Ayo this command doesn't exist!", color = self.client.color)
                embed.description = 'Did you made a typo, or are you curious as to if this command exists? Try again!'
        if normalCommand:
            await ctx.send(embed = embed, reference = ctx.message)
        else:
            await ctx.send(embed = embed)

    async def changelog(self, ctx, normalCommand):
        embed = discord.Embed(title = 'Changelog', color = self.client.color)
        embed.description = "This command has been moved to [Github](https://github.com/BugGlitchy64/ServerKit/blob/master/CHANGELOG.md)"
        if normalCommand:
            await ctx.send(embed = embed, reference = ctx.message)
        else:
            await ctx.send(embed = embed)

    async def ping(self, ctx, normalCommand):
        ping = str(round(self.client.latency * 1000))
        embed = discord.Embed(title = ':ping_pong: Pong!', description = '**' + ping + '**' + 'ms', color = self.client.color)
        if normalCommand:
            await ctx.send(embed = embed, reference = ctx.message)
        else:
            await ctx.send(embed = embed)
        
    @commands.command(
        name = 'info', description = "Shows the info about the bot!", usage = "Information"
    )
    async def normalInfo(self, ctx):
        await self.info(ctx, True)

    @commands.command(
        name = 'help', aliases = ['commands'], description = "The help command lists all of the commands usable in the bot!", usage = "Information"
    )
    async def normalHelp(self, ctx, arg=None):
        await self.help(ctx, True, arg)

    @commands.command(
        name = 'changelog', description = "Shows the version log about the bot!", usage = "Information"
    )
    async def normalChangelog(self, ctx):
        await self.changelog(ctx, True)

    @commands.command(
        name = 'ping', description = "Measure the delay between you and the bot!", usage = "Information"
    )
    async def normalPing(self, ctx):
        await self.ping(ctx, True)

    @cog_ext.cog_slash(
        name = 'info', description = "Shows the info about the bot!"
    )
    async def slashInfo(self, ctx):
        await self.info(ctx, False)

    @cog_ext.cog_slash(
        name = 'help', description = "The help command lists all of the commands usable in the bot!",
        options=[
            create_option(
                 name="command",
                 description="Insert a command to see descriptions and other info.",
                 option_type=3,
                 required=False
               )
        ]
    )
    async def slashHelp(self, ctx, arg=None):
        await self.help(ctx, False, arg)

    @cog_ext.cog_slash(
        name = 'changelog', description = "Shows the version log about the bot!"
    )
    async def slashChangelog(self, ctx):
        await self.changelog(ctx, False)

    @cog_ext.cog_slash(
        name = 'ping', description = "Measure the delay between you and the bot!"
    )
    async def slashPing(self, ctx):
        await self.ping(ctx, False)

def setup(client):
    client.add_cog(info(client))